crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -1,59 +1,115 @@
+    )]
     MissingRequiredScripts(Vec<ScriptHash>),
-    #[error("extraneous script witnesses: extra {0:?}")]
+    #[error("extraneous script witnesses: extra [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     ExtraneousScriptWitnesses(Vec<ScriptHash>),
+    #[error("unspendable inputs; no datums: [{}]",
+        .0
+        .iter()
+        .map(|input|
+            format!("{}#{}", input.transaction_id, input.index)
+        )
+        .collect::<Vec<_>>()
+        .join(", ")
+    )]
+    UnspendableInputsNoDatums(Vec<TransactionInput>),
+    #[error(
+        "missing required datums: missing [{}] provided [{}]",
+        missing.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    MissingRequiredDatums {
+        missing: Vec<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
+    #[error(
+        "extraneous supplemental datums: allowed: [{}] provided [{}]",
+        allowed.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    ExtraneousSupplementalDatums {
+        allowed: BTreeSet<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
 }
 
+// TODO: this can be made MUCH more efficient. Remove clones, don't iterate the same list several times, etc... Lots of low hanging fruit.
 pub fn execute<C>(
     context: &mut C,
     reference_inputs: Option<&Vec<TransactionInput>>,
-    inputs: &Vec<TransactionInput>,
+    inputs: &[TransactionInput],
     witness_set: &MintedWitnessSet<'_>,
 ) -> Result<(), InvalidScripts>
 where
     C: UtxoSlice + WitnessSlice,
 {
     let required_scripts = context.required_scripts();
 
+    let resolved_inputs = inputs
+        .iter()
+        .filter_map(|input| {
+            context
+                // We assume that the input exists as that's validated during the inputs validation
+                .lookup(input)
+                .map(|output| (input, output))
+        })
+        .collect::<Vec<_>>();
+
+    let empty_vec = vec![];
+    let resolved_reference_inputs = reference_inputs
+        .unwrap_or(&empty_vec)
+        .iter()
+        .filter_map(|input| {
+            context
+                // We assume that the reference input exists as that's validated during the inputs validation
+                .lookup(input)
+                .map(|output| (input, output))
+        })
+        .collect::<Vec<_>>();
+
     // provided reference scripts from inputs and reference inputs only include ScriptRefs that are required by an input
     let provided_reference_scripts = [
-        reference_inputs.unwrap_or(&vec![]).as_slice(),
-        inputs.as_slice(),
+        resolved_inputs.as_slice(),
+        resolved_reference_inputs.as_slice(),
     ]
     .concat()
```

Comment:
Looks like some intermediary `collect` and `concat` (allocation in general) could be avoided by keeping iterators alive longer?

Reply:
Almost certainly. I'm working through a set of validations that will make up this `scripts.rs` in separate PRs, then I will go back through and cleanup/improve this. My first focus is just getting *something* working with some tests so that I can more effectively optimize.